### PR TITLE
Update absl

### PR DESCRIPTION
**Pull request by snnn**
### Motivation and Context
Our macOS pipeline are failing because of a build error in absl. However, the bug fix we need is not available in the latest ABSL release. 

Here  is the issue: https://github.com/abseil/abseil-cpp/pull/1536
And here is the fix: https://github.com/abseil/abseil-cpp/commit/779a3565ac6c5b69dd1ab9183e500a27633117d5
 
GTests uses ABSL. But this ABSL target also depends on GTest. So, it is a circular dependency. We should be able to avoid that by avoid building tests for ABSL. However, the version we are using has a problem with that: it has cmake target that still depends on GTest even when testing is disabled. 

I don't know why we suddenly hit this problem.

### CODE REVIEW

**Comment by yuslepukhin** [comment]
Do we know if the internals of the classes are changing? Question is related to natvis.

**Comment by snnn** [comment]
Most likely they are changing.

**Comment by snnn** [comment]
GCC 12 seems to have a bug with the latest ABSL code.
```
$ /opt/rh/gcc-toolset-12/root/usr/bin/c++ -DCPUINFO_SUPPORTED_PLATFORM=1 -DEIGEN_MPL2_ONLY -DEIGEN_USE_THREADS -DNSYNC_ATOMIC_CPP11 -DONLY_C_LOCALE=0 -DONNX_ML=1 -DONNX_NAMESPACE=onnx -DONNX_USE_LITE_PROTO=1 -DORT_ENABLE_STREAM -DORT_RUN_EXTERNAL_ONNX_TESTS -DPLATFORM_POSIX -D_GNU_SOURCE -D__ONNX_NO_DOC_STRINGS -I/tmp/build5/Release/_deps/utf8_range-src -I/onnxruntime_src/include/onnxruntime -I/onnxruntime_src/include/onnxruntime/core/session -I/tmp/build5/Release/_deps/pytorch_cpuinfo-src/include -I/tmp/build5/Release/_deps/google_nsync-src/public -I/tmp/build5/Release -I/onnxruntime_src/onnxruntime -I/tmp/build5/Release/_deps/abseil_cpp-src -I/tmp/build5/Release/_deps/safeint-src -I/tmp/build5/Release/_deps/gsl-src/include -I/tmp/build5/Release/_deps/date-src/include -I/tmp/build5/Release/_deps/onnx-src -I/tmp/build5/Release/_deps/onnx-build -I/tmp/build5/Release/_deps/protobuf-src/src -I/tmp/build5/Release/_deps/flatbuffers-src/include -I/tmp/build5/Release/_deps/mp11-src/include -I/tmp/build5/Release/_deps/eigen-src -I/tmp/build5/Release/_deps/re2-src -DNDEBUG -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -O3 -pipe -fstack-clash-protection -fcf-protection -ffunction-sections -fdata-sections -Wno-restrict  -DCPUINFO_SUPPORTED -O3 -DNDEBUG -fPIC -Wall -Wextra -Wno-deprecated-copy -Wno-nonnull-compare -Werror /onnxruntime_src/slice.cc  -c
In file included from /tmp/build5/Release/_deps/abseil_cpp-src/absl/container/flat_hash_set.h:45,
                 from /onnxruntime_src/include/onnxruntime/core/common/inlined_containers.h:21,
                 from /onnxruntime_src/include/onnxruntime/core/graph/graph.h:23,
                 from /onnxruntime_src/include/onnxruntime/core/graph/graph_viewer.h:8,
                 from /onnxruntime_src/include/onnxruntime/core/framework/op_node_proto_helper.h:9,
                 from /onnxruntime_src/include/onnxruntime/core/framework/op_kernel_info.h:9,
                 from /onnxruntime_src/include/onnxruntime/core/framework/op_kernel.h:20,
                 from /onnxruntime_src/onnxruntime/core/providers/cpu/tensor/slice.h:7,
                 from /onnxruntime_src/slice.cc:4:
/tmp/build5/Release/_deps/abseil_cpp-src/absl/container/internal/raw_hash_set.h: In function ‘onnxruntime::common::Status onnxruntime::SliceOp::PrepareForComputeHelper(const gsl::span<const long int>&, const gsl::span<const long int>&, const gsl::span<const long int>&, const gsl::span<const long int>&, PrepareForComputeMetadata&)’:
/tmp/build5/Release/_deps/abseil_cpp-src/absl/container/internal/raw_hash_set.h:1292:54: error: ‘unique_axes.onnxruntime::InlinedHashSet<long int, std::allocator<long int> >::<unnamed>.absl::flat_hash_set<long int, absl::hash_internal::Hash<long int>, std::equal_to<long int>, std::allocator<long int> >::<unnamed>.absl::container_internal::raw_hash_set<absl::container_internal::FlatHashSetPolicy<long int>, absl::hash_internal::Hash<long int>, std::equal_to<long int>, std::allocator<long int> >::settings_.absl::container_internal::CompressedTuple<absl::container_internal::CommonFields, absl::hash_internal::Hash<long int>, std::equal_to<long int>, std::allocator<long int> >::<unnamed>.absl::container_internal::internal_compressed_tuple::CompressedTupleImpl<absl::container_internal::CompressedTuple<absl::container_internal::CommonFields, absl::hash_internal::Hash<long int>, std::equal_to<long int>, std::allocator<long int> >, std::integer_sequence<long unsigned int, 0, 1, 2, 3>, true>::<unnamed>.absl::container_internal::internal_compressed_tuple::Storage<absl::container_internal::CommonFields, 0, false>::value.absl::container_internal::CommonFields::heap_or_soo_.absl::container_internal::HeapOrSoo::heap.absl::container_internal::HeapPtrs::control’ may be used uninitialized [-Werror=maybe-uninitialized]
 1292 |     ABSL_SWISSTABLE_IGNORE_UNINITIALIZED_RETURN(heap.control);
      |                                                      ^~~~~~~
In file included from /onnxruntime_src/slice.cc:13:
/onnxruntime_src/onnxruntime/core/providers/cpu/tensor/slice_helper.h:98:11: note: ‘unique_axes’ declared here
   98 |   AxesSet unique_axes;
      |           ^~~~~~~~~~~
/tmp/build5/Release/_deps/abseil_cpp-src/absl/container/internal/raw_hash_set.h: In function ‘onnxruntime::common::Status onnxruntime::SliceOp::PrepareForComputeHelper(const gsl::span<const long int>&, const gsl::span<const long int>&, const gsl::span<const long int>&, PrepareForComputeMetadata&)’:
/tmp/build5/Release/_deps/abseil_cpp-src/absl/container/internal/raw_hash_set.h:1292:54: error: ‘unique_axes.onnxruntime::InlinedHashSet<long int, std::allocator<long int> >::<unnamed>.absl::flat_hash_set<long int, absl::hash_internal::Hash<long int>, std::equal_to<long int>, std::allocator<long int> >::<unnamed>.absl::container_internal::raw_hash_set<absl::container_internal::FlatHashSetPolicy<long int>, absl::hash_internal::Hash<long int>, std::equal_to<long int>, std::allocator<long int> >::settings_.absl::container_internal::CompressedTuple<absl::container_internal::CommonFields, absl::hash_internal::Hash<long int>, std::equal_to<long int>, std::allocator<long int> >::<unnamed>.absl::container_internal::internal_compressed_tuple::CompressedTupleImpl<absl::container_internal::CompressedTuple<absl::container_internal::CommonFields, absl::hash_internal::Hash<long int>, std::equal_to<long int>, std::allocator<long int> >, std::integer_sequence<long unsigned int, 0, 1, 2, 3>, true>::<unnamed>.absl::container_internal::internal_compressed_tuple::Storage<absl::container_internal::CommonFields, 0, false>::value.absl::container_internal::CommonFields::heap_or_soo_.absl::container_internal::HeapOrSoo::heap.absl::container_internal::HeapPtrs::control’ may be used uninitialized [-Werror=maybe-uninitialized]
 1292 |     ABSL_SWISSTABLE_IGNORE_UNINITIALIZED_RETURN(heap.control);
      |                                                      ^~~~~~~
/onnxruntime_src/onnxruntime/core/providers/cpu/tensor/slice_helper.h:39:11: note: ‘unique_axes’ declared here
   39 |   AxesSet unique_axes;
      |           ^~~~~~~~~~~
cc1plus: all warnings being treated as errors
```

GCC 13 doesn't have this issue.

File: onnxruntime/core/providers/qnn/builder/qnn_utils.cc [hunk 1]
```diff
@@ -319,6 +319,8 @@ std::ostream& operator<<(std::ostream& out, const Qnn_Tensor_t& tensor) {
   }
   out << ")";
   out << " memType=" << GetQnnTensorMemType(tensor);
+// TODO: the code below has compilation errors with the latest ABSL
```
Comment (jywu-msft) [inline onnxruntime/core/providers/qnn/builder/qnn_utils.cc:322]: +@hectorsvc @adrianlizarraga FYI , is this ok?

Comment (snnn) [reply onnxruntime/core/providers/qnn/builder/qnn_utils.cc:322]: The build error was:
```
In file included from /usr/include/c++/11/random:49,
                 from /mnt/vss/_work/1/s/build/Release/_deps/abseil_cpp-src/absl/algorithm/container.h:47,
                 from /mnt/vss/_work/1/s/build/Release/_deps/abseil_cpp-src/absl/container/flat_hash_set.h:40,
                 from /mnt/vss/_work/1/s/include/onnxruntime/core/common/inlined_containers.h:21,
                 from /mnt/vss/_work/1/s/include/onnxruntime/core/graph/graph.h:23,
                 from /mnt/vss/_work/1/s/onnxruntime/core/framework/node_unit.h:15,
                 from /mnt/vss/_work/1/s/onnxruntime/core/providers/qnn/builder/qnn_utils.h:13,
                 from /mnt/vss/_work/1/s/onnxruntime/core/providers/qnn/builder/qnn_utils.cc:12:
/usr/include/c++/11/bits/random.h: In instantiation of ‘class std::discrete_distribution<float>’:
/mnt/vss/_work/1/s/onnxruntime/core/providers/qnn/builder/qnn_utils.cc:324:25:   required from here
/usr/include/c++/11/bits/random.h:5281:49: error: static assertion failed: result_type must be an integral type
 5281 |       static_assert(std::is_integral<_IntType>::value,
      |                                                 ^~~~~
/usr/include/c++/11/bits/random.h:5281:49: note: ‘std::integral_constant<bool, false>::value’ evaluates to false
/usr/include/c++/11/bits/random.h: In instantiation of ‘class std::binomial_distribution<float>’:

```

And the build command was:
/usr/bin/c++ -DCPUINFO_SUPPORTED_PLATFORM=1 -DEIGEN_MPL2_ONLY -DEIGEN_USE_THREADS -DNSYNC_ATOMIC_CPP11 -DONLY_C_LOCALE=0 -DONNX_ML=1 -DONNX_NAMESPACE=onnx -DONNX_USE_LITE_PROTO=1 -DORT_ENABLE_STREAM -DPLATFORM_POSIX -DUSE_QNN=1 -D_GNU_SOURCE -D__ONNX_NO_DOC_STRINGS -I/mnt/vss/_work/1/s/build/Release/_deps/utf8_range-src -I/mnt/vss/_work/1/s/include/onnxruntime -I/mnt/vss/_work/1/s/include/onnxruntime/core/session -I/mnt/vss/_work/1/s/build/Release/_deps/pytorch_cpuinfo-src/include -I/mnt/vss/_work/1/s/build/Release/_deps/google_nsync-src/public -I/mnt/vss/_work/1/s/build/Release -I/mnt/vss/_work/1/s/onnxruntime -I/mnt/vss/_work/1/s/build/Release/_deps/abseil_cpp-src -I/mnt/vss/_work/1/s/build/Release/_deps/safeint-src -I/mnt/vss/_work/1/s/build/Release/_deps/gsl-src/include -I/mnt/vss/_work/1/s/build/Release/_deps/date-src/include -I/mnt/vss/_work/1/s/build/Release/_deps/onnx-src -I/mnt/vss/_work/1/s/build/Release/_deps/onnx-build -I/mnt/vss/_work/1/s/build/Release/_deps/protobuf-src/src -I/mnt/vss/_work/1/s/build/Release/_deps/flatbuffers-src/include -I/mnt/vss/_work/1/s/build/Release/_deps/mp11-src/include -I/mnt/vss/_work/_temp/qnn-v2.23.0.240531/include/QNN -I/mnt/vss/_work/_temp/qnn-v2.23.0.240531/include -DNDEBUG -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -O3 -pipe -fstack-clash-protection -fcf-protection -ffunction-sections -fdata-sections -Wno-restrict  -DCPUINFO_SUPPORTED -O3 -DNDEBUG -std=gnu++17 -fPIC -Wall -Wextra -Wno-deprecated-copy -Wno-nonnull-compare -Wno-unknown-pragmas -Werror -MD -MT CMakeFiles/onnxruntime_providers_qnn.dir/mnt/vss/_work/1/s/onnxruntime/core/providers/qnn/builder/qnn_utils.cc.o -MF CMakeFiles/onnxruntime_providers_qnn.dir/mnt/vss/_work/1/s/onnxruntime/core/providers/qnn/builder/qnn_utils.cc.o.d -o CMakeFiles/onnxruntime_providers_qnn.dir/mnt/vss/_work/1/s/onnxruntime/core/providers/qnn/builder/qnn_utils.cc.o -c /mnt/vss/_work/1/s/onnxruntime/core/providers/qnn/builder/qnn_utils.cc

Comment (HectorSVC) [reply onnxruntime/core/providers/qnn/builder/qnn_utils.cc:322]: It's OK, the code here just prints information for debugging purposes.

Comment (jywu-msft) [reply onnxruntime/core/providers/qnn/builder/qnn_utils.cc:322]: we should try to fix it later.

**Review by pranavsharma — APPROVED** [review]

**Review by jchen351 — APPROVED** [review]

**Review by HectorSVC — APPROVED** [review]

File: cmake/CMakeLists.txt [hunk 2]
```diff
@@ -1057,6 +1057,9 @@ function(onnxruntime_set_compile_flags target_name)
     foreach(FLAG ${ORT_WARNING_FLAGS})
       target_compile_options(${target_name} PRIVATE "$<$<COMPILE_LANGUAGE:CXX>:${FLAG}>")
     endforeach()
+    if("${CMAKE_C_COMPILER_ID}" STREQUAL "GNU" AND CMAKE_C_COMPILER_VERSION VERSION_LESS 13 AND CMAKE_C_COMPILER_VERSION VERSION_GREATER_EQUAL 12)
+      target_compile_options(${target_name} PRIVATE "$<$<COMPILE_LANGUAGE:CXX>:-Wno-maybe-uninitialized>")
```
Comment (edgchen1) [inline cmake/CMakeLists.txt:1061]: can this disabled warning be localized? e.g., if it's only from absl/container/flat_hash_set.h, and we only include that header through include/onnxruntime/core/common/inlined_containers.h, can we just disable it around that include in inlined_containers.h?

Comment (snnn) [reply cmake/CMakeLists.txt:1061]: This is temporary.  Soon I will upgrade GCC to 13, then we don't need to worry about it.

Comment (snnn) [reply cmake/CMakeLists.txt:1061]: Though the log says the error is from inlined_containers.h, but actually it's hard to find out which line of code caused the issue. I tried to compile the following code alone and it didn't cause any trouble. 
```
AxesSet unique_axes
```
Then I realized it was probably just an internal compiler bug.

**Pull request merged**